Refactor progress loading and saving for signed-out users

jamescodeorg opened the pull request:
<!--
  A summary of the change, including any relevant background, motivation, and context.
  If relevant, include a description, screenshots, and/or video of the existing and new behavior.
-->

The goal of this PR is to consolidate and (hopefully) simplify how we load and save progress for the level page and fix several bugs in the process.  We are specifically focused on the following aspects of progress:
- The TestResult of each level.  This is reflected in the progress bubbles at the top of the level page.
- The student's source code/answer, often referred to as `lastAttempt` in the code.

**Loading Progress**

How level progress is loaded depends on several factors:
- **What type of level it is.**  In particular, a bunch of the logic was in loadApp.js which is specific to lab-based projects.
- **Whether the user is signed in or not.**  Progress for signed-in users is stored in the database and progress for signed-out users is stored in the browser's session storage.
- **Whether the level is a [cached HoC level](https://github.com/code-dot-org/code-dot-org/blob/b0272a6ef85ed7d788469a51e402ffbd39a9b4de/cookbooks/cdo-varnish/libraries/http_cache.rb#L27-L49).**  For regular, non-cached levels, progress is embedded in the html page.  For cached levels, progress is requested from the server asynchronously.

For more details, see [this comment](https://codedotorg.atlassian.net/browse/LP-1810?focusedCommentId=18142) in the Jira ticket.

This PR attempts to unify the codepath for all the cases above, collecting code from several locations and consolidating it in progress.generateStageProgress().  After trying several approaches, I decided to (1) process all cases asynchronously, (2) continue using the progress embedded in the html page when it is available, and (3) request the data from the server when it is not available.  (I considered (and tried!) making all cases load progress asynchronously from the server but this was riskier than I was comfortable with.)

**Saving Progress**

Prior to this change, there were two very different code paths for saving progress to the server and for saving progress locally:
- Progress was saved to the server via reporting.sendReport() which sent a milestone request to the server.
- Progress was saved locally via a listener that listened for changes to the redux store.

I've unified these code paths as follows:
- All level types report progress by calling reporting.sendReport() which does the following:
    - Update redux
    - Saves the report locally if appropriate
    - Sends a milestone report to the server.  The report is always sent; the server decides whether to save it to the database depending on if the user is logged in.


## Links

<!--
  Links to relevant external resources; ie, specification documents, Jira tickets, related PRs, Honeybadger errors, etc.
-->

[LP-1564](https://codedotorg.atlassian.net/browse/LP-1564)
[LP-1810](https://codedotorg.atlassian.net/browse/LP-1810)
[LP-1822](https://codedotorg.atlassian.net/browse/LP-1822)

## Testing story

<!--
  Does your change include appropriate tests?
  If so, please describe how the tests included in this PR are sufficient.
  If not, please explain why this change does not need to be tested.
-->

There were existing UI tests for verifying level progress for cached (HoC) levels in hour_of_code.feature and hour_of_code_signed_in.feature.  I've added new UI tests for verify level progress for regular uncached levels.

I'm deferring unit tests for the js code because the code is currently structured in a way that makes unit testing difficult and I'm anticipating larger, structural changes to this code in the near future.

The team has also done a bug bash.

<!-- Other aspects to consider. Delete any sections that are not relevant to your change. -->

## Deployment strategy

## Follow-up work

<!--
  List (ideally with Jira links) any clean-up or technical debt that will be addressed in future work.
-->

## Privacy

<!--
  1.	Does this change involve the collection, use, or sharing of new Personal Data?
  2.	Does this change involve a new or changed use or sharing of existing Personal Data?
-->

## Security

<!-- Link to Jira task(s) where sensitive security issues are discussed privately. -->

## Caching

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

cforkish commented on apps/src/code-studio/clientState.js:
```diff
@@ -37,9 +38,13 @@ clientState.reset = function() {
 clientState.clearProgress = function() {
   sessionStorage.removeItem('progress');
   sessionStorage.removeItem('lines');
+  removeItemsWithPrefix(sessionStorage, 'source_');
```
can you comment on this change? explaining this line in the header comment might be nice too. the first two lines are pretty clear, but this one is not self-explanatory

cforkish replied:
never mind. the `sourceForLevel` function below was collapsed, but seeing it there for context make this more self-explanatory

cforkish commented on apps/src/code-studio/initApp/loadApp.js:
```diff
@@ -399,68 +307,33 @@ function loadAppAsync(appOptions) {
-            appOptions.level.pairingChannelId = data.pairingChannelId;
-          }
+        // We do not need to process data.progress here because labs do not use
+        // the the level progress data directly. (The progress bubbles in the
```
typo

cforkish commented on apps/src/code-studio/progress.js:
```diff
@@ -234,12 +304,18 @@ function queryUserProgress(store, scriptData, currentLevelId) {
+    // If the user is not signed in, retrieve level progress from session storage.
+    // (If the user is signed in, level progress would have been set by the call
+    // to reduxQueryUserProgress above.)
+    if (!data.signedIn) {
```
should we also check that user_id === null?

jamescodeorg replied:
Actually, after looking at this some more, I think we probably don't need to check that `user_id === null` here.  This is called from the course overview page and user_id should only be set when the teacher is viewing a student which only applies when the teacher is signed in.  If the user is not signed in, the code should generally be ignoring the user_id parameter even if it is provided (because an anonymous user should not be able to request progress for an arbitrary user).

cforkish replied:
👍

hannahbergam commented on apps/src/code-studio/header.js:
```diff
@@ -122,7 +116,7 @@ header.build = function(
     // Only render sign in callout if the course is CSF and the user is
     // not signed in
-    if (scriptData.is_csf && !signedIn) {
+    if (scriptData.is_csf && signedIn === false) {
```
Just for my own curiosity- are there other valid states of signedIn that are not true/false? Or is this just a clarity preference?

hannahbergam commented on apps/src/code-studio/progress.js:
```diff
@@ -120,26 +119,96 @@ progress.generateStageProgress = function(
+        usingDbProgress: false,
+        progress: clientState.levelProgress(scriptName)
+      });
+    case null:
```
(followup- I see! disregard my question about signedIn values)

hannahbergam reviewed (commented):
This looks like a huge improvement and consolidation. Nice! My only thought- I know we had a pretty stellar bug bash on this, but are there any ways to write in test cases for the bugs you fixed along the way?

jamescodeorg commented:
@hannahbergam Thanks for the reminder about the tests.  I verified that we have existing UI tests that verify progress for the cached+signed-in and cached+signed out cases and added two new tests to cover uncached+signed-in and uncached+signed-out cases.  I looked at adding some unit tests for loadApp.js but I think unit testing would be easier/more effective if we refactored the component and want to defer that for now.  Let me know if you disagree.

hannahbergam commented:
> @hannahbergam Thanks for the reminder about the tests. I verified that we have existing UI tests that verify progress for the cached+signed-in and cached+signed out cases and added two new tests to cover uncached+signed-in and uncached+signed-out cases. I looked at adding some unit tests for loadApp.js but I think unit testing would be easier/more effective if we refactored the component and want to defer that for now. Let me know if you disagree.

I'm happy with UI tests being the solid fallback, as I've learned they sometimes fill the role of integration tests for us:) Looks good to me!

cforkish commented on apps/src/code-studio/progress.js:
```diff
@@ -125,7 +126,21 @@ progress.generateStageProgress = function(
   }
 
-  return getLevelProgress(signedIn, progressData, name).then(data => {
+  return populateProgress(store, signedIn, progressData, name);
```
:)

cforkish reviewed (approved):
LGTM! so much work for such small changes. ;-)
